File: Exp 10 Matrix Chain Multiplication using Dynamic Programming Approach/chain_matrix_multiplication.cpp
#include <iostream>
using namespace std;

const int MAXN = 10; // maximum number of matrices

// print optimal parenthesization using split table s
void printParens(int s[][MAXN], int i, int j) {
    if (i == j) {
        cout << "A" << i;
        return;
    }
    cout << "(";
    printParens(s, i, s[i][j]);
    printParens(s, s[i][j] + 1, j);
    cout << ")";
}

// matrix chain order: p has length n+1 (p[0..n])
// returns minimal cost and fills s with split points
long long matrixChainOrder(int p[], int n, long long m[][MAXN], int s[][MAXN]) {
    // m[i][j] will hold minimal cost for multiplying Ai..Aj (1-based indices)
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= n; ++j) {
            if (i == j) m[i][j] = 0;
            else m[i][j] = (long long)1e18; // INF
            s[i][j] = 0;
        }
    }

    // L is chain length
    for (int L = 2; L <= n; ++L) {
        for (int i = 1; i <= n - L + 1; ++i) {
            int j = i + L - 1;
            // try all splits
            for (int k = i; k <= j - 1; ++k) {
                long long q = m[i][k] + m[k + 1][j] + (long long)p[i - 1] * p[k] * p[j];
                if (q < m[i][j]) {
                    m[i][j] = q;
                    s[i][j] = k;
                }
            }
        }
    }

    return m[1][n];
}

int main() {
    int n;
    cout << "Enter number of matrices (n, max " << MAXN-1 << "): ";
    cin >> n;
    if (n <= 0 || n >= MAXN) {
        cout << "Invalid n (1.." << (MAXN-1) << ")\n";
        return 0;
    }

    int p[MAXN];
    cout << "Enter " << (n + 1) << " dimensions p0..p" << n << " (space separated):\n";
    for (int i = 0; i <= n; ++i) cin >> p[i];

    long long m[MAXN][MAXN];
    int s[MAXN][MAXN];

    long long minCost = matrixChainOrder(p, n, m, s);

    cout << "\nMinimum number of scalar multiplications = " << minCost << "\n";
    cout << "Optimal parenthesization: ";
    printParens(s, 1, n);
    cout << "\n";

    return 0;
}

File: Exp 10 Matrix Chain Multiplication using Dynamic Programming Approach/chain_matrix_multiplication.py
MAXN = 10  # maximum number of matrices


def print_parens(s, i, j):
    """Print optimal parenthesization using split table s"""
    if i == j:
        print(f"A{i}", end="")
        return
    print("(", end="")
    print_parens(s, i, s[i][j])
    print_parens(s, s[i][j] + 1, j)
    print(")", end="")


def matrix_chain_order(p, n):
    """
    Matrix chain order: p has length n+1 (p[0..n])
    Returns minimal cost and fills s with split points
    """
    # m[i][j] will hold minimal cost for multiplying Ai..Aj (1-based indices)
    m = [[0] * (MAXN + 1) for _ in range(MAXN + 1)]
    s = [[0] * (MAXN + 1) for _ in range(MAXN + 1)]
    
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if i == j:
                m[i][j] = 0
            else:
                m[i][j] = float('inf')  # INF
            s[i][j] = 0
    
    # L is chain length
    for L in range(2, n + 1):
        for i in range(1, n - L + 2):
            j = i + L - 1
            # Try all splits
            for k in range(i, j):
                q = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j]
                if q < m[i][j]:
                    m[i][j] = q
                    s[i][j] = k
    
    return m[1][n], m, s


def main():
    n = int(input(f"Enter number of matrices (n, max {MAXN - 1}): "))
    if n <= 0 or n >= MAXN:
        print(f"Invalid n (1..{MAXN - 1})")
        return
    
    p = [0] * MAXN
    print(f"Enter {n + 1} dimensions p0..p{n} (space separated):")
    dimensions = list(map(int, input().split()))
    for i in range(n + 1):
        p[i] = dimensions[i]
    
    min_cost, m, s = matrix_chain_order(p, n)
    
    print(f"\nMinimum number of scalar multiplications = {min_cost}")
    print("Optimal parenthesization: ", end="")
    print_parens(s, 1, n)
    print()


if __name__ == "__main__":
    main()
